src/grav/test.py

```diff
@@ -1,116 +1,141 @@
-    json = resp.json()
-    match = re.match(r"Loaded image ID: (sha256:[\da-f]{64})", json['stream'])
-    image_id = match.group(1)
-
-    # create container, get container id
-    data = {"Image": image_id}
-    headers = {"Content-Type": "application/json"}
-    resp = client.post("/containers/create", body=data, headers=headers)
-    assert resp.status == 201
-    json = resp.json()
-    container_id = json['Id']
-
-    return client, image_id, container_id
-
-
-def test(client, container_id):
-    # start
-    start_path = "/containers/%s/start" % container_id
-    resp = client.post(start_path)
-    resp.read()  # TODO: have this handled by the client
-    assert resp.status == 204
-
-    # stop
-    stop_path = "/containers/%s/stop" % container_id
-    resp = client.post(stop_path)
-    resp.read()  # TODO: have this handled by the client
-    assert resp.status == 204
-
-    # logs
-    logs_path = "/containers/%s/logs" % container_id
-    logs_path += "?stdout=true&stderr=true"
-    resp = client.get(logs_path)
-    assert resp.status == 200
-    # logs doesn't return json, but a multiplexed stream
-    stdout, stderr = demux_logs(resp.read())
-
-    # validate
-    lines = stdout.splitlines()
-    assert lines[0] == "Hello Gravitational!"
-    assert lines[-1] == "Terminated. Bye!"
-    assert stderr == ""
-
-
-def teardown(client, container_id, image_id):
-    # if the test dies partway through, the container may
-    # still be running, and should be stopped
-    inspect_path = "/containers/%s/json" % container_id
-    resp = client.get(inspect_path)
-    assert resp.status == 200
-    json = resp.json()
-    state = json["State"]
-    running = state["Running"]
-    if running:
         # stop
         stop_path = "/containers/%s/stop" % container_id
         resp = client.post(stop_path)
         resp.read()  # TODO: have this handled by the client
-        assert resp.status == 204
-
-    # delete container
-    delete_path = "/containers/" + container_id
-    resp = client.request("DELETE", delete_path)
-    resp.read()  # TODO: have this handled by the client
-    assert resp.status == 204
-
-    # delete image
-    delete_path = "/images/" + image_id
-    resp = client.request("DELETE", delete_path)
-    resp.read()  # TODO: have this handled by the client
-    assert resp.status == 200  # why isn't this 204 like the other api endpoints?
-
+        self.assertStatus(resp, 204)
+
+        # logs
+        logs_path = "/containers/%s/logs" % container_id
+        logs_path += "?stdout=true&stderr=true"
+        resp = client.get(logs_path)
+        self.assertStatus(resp, 200)
+        # logs doesn't return json, but a multiplexed stream
+        stdout, stderr = demux_logs(resp.read())
+
+        # validate
+        stdout_lines = stdout.splitlines()
+        self.assertEqual(stdout_lines[0], "Hello Gravitational!")
+        self.assertEqual(stdout_lines[-1], "Terminated. Bye!")
+        self.assertEqual(stderr, "")
```

Comment:
Better to assert that stderr is empty before checking stdout. Incase of err could stdout be empty?

Reply:
Good point. We'll get better error messages with this suggestion.

Reply:
Addressed by 9a3cd96.